[Haikal Yusuf] iP

waffledood opened the pull request:
# Duke 🤖
>“Deciding what NOT to do is as important as deciding what to do.”
— Steve Jobs [(source)](https://quotefancy.com/quote/267711/Steve-Jobs-Deciding-what-not-to-do-is-as-important-as-deciding-what-to-do)

Duke helps you to remember the things you need to do! It is
- Text based 📝
- Easy to learn 👶 
- ~~FAST~~ SUPER FAST to use 🏃🏻💨
- FREE! 🤩 

# Getting Started 🤔
1. Download the latest release from [here](https://github.com/waffledood/ip/releases/tag/A-Jar).
2. Launch using the following command `java- jar ip.jar`. 
3. Add your tasks. 
4. Duke will manage your tasks! 🚀 

# Using Duke 🤠
Duke can perform different functions by giving it the relevant command. Below is a guide. 
- Add a Todo Task → `todo buy milk`
- Add an Event Task → `event Sam's New Year's party /at 2021-01-01` 
- Add a Deadline Task → `deadline return library book /by 2021-02-02` 
- Mark a Task → `mark 3` 
- Unmark a Task → `unmark 3` 
- List Tasks → `list` 
- Exit Duke → `bye` 
- Search for a Task → `find milk`

The types of Commands `Duke` can perform are defined in `CommandType.java`. `Duke` can recognize or perform new commands based on the `regex` parameter. `null` indicates the command does not have an accompanying regex, while commands such as `Event` require the inclusion of an `/at` key. 
```java
public enum CommandType {
    BYE("bye", null),
    LIST("list", null),
    MARK("mark", "\\d+"),
    UNMARK("unmark", "\\d+"),
    DELETE("delete", "\\d+"),
    FIND("find", null),
    TODO("todo", null),
    EVENT("event", "/at"),
    DEADLINE("deadline", "/by");
}
```

# Features 🚀
- [X] Managing different tasks (Todo's, Events, Deadlines) 
- [X] Data writing to & loading from disc 
- [X] Searching for tasks by description 
- [X] Reminders

Commit:
Add Gradle support

waffledood committed:
Level-1

waffledood committed:
Level-2

waffledood committed:
Level-3

waffledood committed:
Level-4

waffledood committed:
Added sample user input & expected output

waffledood committed:
Implemented regex that fixes #3

waffledood committed:
Created DukeException to handle errors specific to Duke

waffledood committed:
Throw DukeException to handle invalid commands, this fixes #4

waffledood committed:
Check for invalid task index, this fixes #6

waffledood committed:
Resolved incomplete commands, fixes #5

waffledood committed:
applied trim() to user input, fixes #8

waffledood committed:
Shifted code for successful add of task, fixes #9

waffledood committed:
Implemented delete feature

waffledood committed:
Corrected mark/unmark feature to recognize inputs of "mark/unmark" only

waffledood committed:
Refactoring Duke logic to use switch

waffledood committed:
Added Command as Enums

waffledood committed:
Integrated Command Enums with Duke

waffledood committed:
Refactor Task types to accept boolean as argument

waffledood committed:
Implement load & saving of data

waffledood committed:
Format Deadline  & Event to recognize dates as LocalDate instead of String

waffledood committed:
Refactor Duke to accommodate LocalDate format of Event & Deadline tasks

waffledood committed:
Merge branch 'branch-Level-7'

waffledood committed:
Merge branch 'branch-Level-8'

waffledood committed:
Fix bug of Duke writing wrong format of time into save file

waffledood committed:
Modify error messages to be printed to the error output, stderr

waffledood committed:
Extract UI from Duke as its own class

waffledood committed:
Refactor Command to handle invalid user input of command internally

waffledood committed:
Extract loading & saving of tasks to file as Storage class

waffledood committed:
Rename Command enums to CommandType

waffledood committed:
Implement Command as an asbtract class & extract commands as standalone classes

waffledood committed:
Extract TaskList as its own class to handle task operations

waffledood committed:
Add isEmpty() method for TaskList

waffledood committed:
Extract various commands' logic to respective command classes

waffledood committed:
Implement Parser.java to read & extract user input

waffledood committed:
Incorrect String comparison fixes #15 bug

waffledood committed:
Organize classes into packages

waffledood committed:
Reorganize files, fix previous organization

waffledood committed:
Add JUnit tests

waffledood committed:
Added Java Docs

waffledood committed:
Resolve bug in Duke where unrecognized commands caused failure, fixes #14

waffledood committed:
Modify switch-case logic in Parser.java, fixes #16

waffledood committed:
Implemented find feature

waffledood committed:
Merge branch 'add-gradle-support'

waffledood committed:
Add Checkstyle

waffledood committed:
Fix checkstyle errors in Duke.java & DukeException.java

waffledood committed:
Fix checkstyle errors

hanqinilnix commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,39 @@
+public abstract class Command {
+
+    protected CommandType commandType;
+    protected boolean active;
```
```suggestion
    protected boolean isActive;
```
Perhaps a more intuitive name here? For example isActive

hanqinilnix commented on src/main/java/duke/command/CommandType.java:
```diff
@@ -0,0 +1,66 @@
+    EVENT("event", "/at"),
+    DEADLINE("deadline", "/by");
+
+    private static final Map<String, CommandType> commandMap = new HashMap<>();
```
```suggestion
    private static final Map<String, CommandType> COMMAND_MAP = new HashMap<>();
```
Please capitalise your constant variables. I noticed that there are similar issues in other parts of your code too

hanqinilnix commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,79 @@
+public class MarkCommand extends Command {
+
+    // mark boolean is true indicates the mark command, false the unmark command
+    private boolean mark;
```
```suggestion
    private boolean isMark;
```
Same as above, more intuitive naming for a Boolean variable. Perhaps there are more of such issues in other files?

hanqinilnix commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDate by;
```
Maybe the variable name can be more descriptive?

hanqinilnix commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,110 @@
+     * @param errorMessage String of errorMessage to be printed.
+     */
+    public void showError(String errorMessage) {
+        System.err.println(String.format("%s%s", indentationText, errorMessage));
```
Cool use .err instead of .out!

hanqinilnix commented on test/duke/task/TaskListTest.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Test
+    void checkSize_add1Task_success() {
+        TaskList taskList = new TaskList(new ArrayList<Task>());
+        taskList.add( new Todo("todo task description", false) );
```
Perhaps you can make these lines of code an attribute of the TaskListTest class?

hanqinilnix commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,104 @@
+
+            // checks if the user already has existing save data
+            if (!f.createNewFile()) {
+                ui.showText("loading previous save data...");
```
I really like this line that informs the user that you are retrieving data.

hanqinilnix reviewed (commented):
Let's do this :)

waffledood committed:
Modify build.gradle to include JavaFX

waffledood committed:
gui folder to store files from JavaFX tutorial

eman-kom commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+     * Setter method to mark the status of a Task as done.
+     */
+    public void markAsDone() {
+        if (this.done == true) {
```
the if statement can be shortened to `if (this.done)`

eman-kom replied:
Can be shortened
```suggestion
        if (this.done) {
```

eman-kom commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,104 @@
+        // write & update save data with current tasklist
+        try {
+            FileWriter fw = new FileWriter(taskFilePath);
+            for (Task t: externalTaskList) {
```
there is no space between the : in the for loop statement

eman-kom replied:
Need to add space between the :
```suggestion
            for (Task t : externalTaskList) {
```

eman-kom commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+     * The main method which intializes Duke & runs it.
+     *
+     * @param args Unused.
+     *             return Nothing.
```
I dont think that the javadoc needs to state that it returns nothing if it returns nothing

eman-kom commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+        userName = ui.showWelcome();
+
+        // initialize the chat session as active
+        boolean active = true;
```
boolean can be written as 'isActive' instead of 'active'
```suggestion
        boolean isActive = true;
```

eman-kom commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,39 @@
+    * Returns a boolean reflecting if the current chat session with Duke is active or not.
+    * @return boolean Boolean indicatinf if the chat session is active.
+    */
+    public abstract boolean isActive();
```
method name should be a verb
```suggestion
    public abstract boolean getActive();
```

eman-kom commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,56 @@
+        boolean tasksMatchingSearchKey = false;
+
+        // cycle through existing tasks & filter tasks
+        for (Task task: taskList.getTasks()) {
```
Need space between :
```suggestion
        for (Task task : taskList.getTasks()) {
```

eman-kom commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,56 @@
+        // if there are tasks that match the searchKey
+        if (tasksMatchingSearchKey) {
+            ui.showText("Here are the matching tasks in your list: ");
+            for (Task task: filteredTaskList) {
```
Need space between :
```suggestion
            for (Task task : filteredTaskList) {
```

eman-kom commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class Event extends Task {
+
+    protected LocalDate at;
```
Variable name can be more descriptive
```suggestion
    protected LocalDate atDateTime;
```

eman-kom commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,104 @@
+            }
+
+        // if the file cannot be read, generate an empty list
+        } catch (IOException e) {
```
Can try to combine exceptions since they are doing the same thing
```suggestion
        } catch (IOException | DukeException e) {
```

eman-kom commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+     * Setter method to mark the status of a Task as undone.
+     */
+    public void markAsUndone() {
+        if (this.done == false) {
```
Can be shortened
```suggestion
        if (!this.done) {
```

eman-kom reviewed (commented):
Great job for following the coding standard 👍 
Just a few minor mistakes

waffledood committed:
Refactor code to achieve basic implementation of GUI

waffledood committed:
Modify AddCommand.java to fix bug

waffledood committed:
Refactor code to improve implementation of Ui

waffledood committed:
Refactor code to improve implementation of Ui

waffledood committed:
Update .gitignore to ignore *.txt files

waffledood committed:
Include custom user picture

waffledood committed:
Merge branch 'Level-10' of https://github.com/waffledood/ip into Level-10

waffledood committed:
Add assertions

waffledood committed:
Clean up Ui.java

waffledood commented:
> Great job for following the coding standard 👍 Just a few minor mistakes

Hey thanks for pointing out all of my mistakes! Some of them I completely overlooked haha 😄 

> Let's do this :)

Thanks for your helpful & critical comments! I appreciate them 😄👍

waffledood committed:
Improve code quality.

The naming of several variables & methods were not in line with the code
quality standard. These were fixed for better readability &
understanding by other devs.

There were also missing javadocs, wrong placements of import statements
that were flagged by the checkstyle.

waffledood committed:
Merge pull request #19 from waffledood/branch-A-CodeQuality

Improve code quality.

waffledood committed:
Implement basic functionality of reminders.

The reminder functionality is achieved by showing an Alert dialog. The
Alert dialog will contain details of the task. The reminder is triggered
when the dates associated with tasks are within 1 day from the time the
application is launched.

waffledood committed:
Reminder implementation with Thread

In this implementation of the Reminder feature, we use a Thread to
cyclically call the lazy function executing the Information alert
dialog popup.

waffledood committed:
Reminder implementation with Timer only, no Thread.

In this implementation of the Reminder feature, the previous
implementation involving the use of Threads was scrapped as the
occurrence of the Thread in line with its sleep rate caused problematic
behavior.

Hence, in this implementation, the start & stop of the Reminder feature
is controlly only via Timer, with the schedule() & cancel() methods.

waffledood committed:
Refactor Task, Deadline, Event & TaskList.

We refactor Task, Deadline & Event so that Task has a LocalDate as one
of its variables so that calling getDate() can be applied to Task.

TaskList is also refactored so that the ArrayList containing the tasks
is a static object. After all, there will only be one instance of the
TaskList in the entire runtime of the application.

waffledood committed:
Implement RemindersTask & ReminderPredicates for Reminder feature.

We implement RemindersTask as an explicit class to store all
functionalities associated with triggering the Reminder function,
building of Reminder alert dialog.

The ReminderPredicates class is implemented to store various Predicates
relating to different types of Reminders that can be set.

waffledood committed:
Update usage of TaskList's getTasks() in relevant classes

waffledood committed:
Create gradle.yml

waffledood committed:
Change pictures of user & duke

waffledood committed:
Update variables in ReminderPredicates.java to be final

waffledood committed:
Refactor GUI for an improve UI.

The MainWindow was set to a fixed size & it won't be resizable.

The overall UI of the chatbot was improved, with the use of additional
styling & color.

waffledood committed:
Remove unused import to fix checkstyle error

waffledood committed:
Merge pull request #20 from waffledood/branch-B-Reminders

Merging Branch B-Reminders

waffledood committed:
Set theme jekyll-theme-minimal

waffledood committed:
Modify DialogBox to dynamically adjust its height

waffledood committed:
Update README.md with relevant info & pictures

waffledood committed:
Fix erroneous text in Ui

waffledood committed:
Include additional Predicates for Task Reminders

waffledood committed:
Set the Task Reminder nature to overdue & upcoming tasks

waffledood committed:
Fix checkstyle errors

waffledood committed:
Add javadocs for these classes.

RemindersTask, ReminderPredicates & DialogBox's javadocs were updated,
as the javadoc comments for the class was missing.

waffledood committed:
Merge pull request #22 from waffledood/branch-A-UserGuide

Branch A-UserGuide

waffledood committed:
Update README.md to match images

waffledood committed:
Update relative path of main Ui picture in README.md

waffledood committed:
Add Ui.png to docs to resolve bug

waffledood committed:
Modify mainClassName in build.gradle to point to Launcher
